jt_notes.md

```diff
@@ -0,0 +1,127 @@
+
+- **IMPORTANT:** the same questions apply to the Flask app. As per my comment in app.py, I am starting to suspect that this could be running concurrent threads. That needs careful thought and documenting, because it gets very complicated.
+
+- More generally, *as far as I can see*, there doesn't seem to be to be a critical need to be using [sequential] threads here at all (though it doesn't do any harm). Out of interest, is this code just adapted from some online example? I suspect it makes a lot more sense in the case where there is *more than one* client involved (but I don't think that's really an issue for us in reality).
+
+**Out of interest:** I find the code formatting [newline after "(", and ")" on a line of its own, sometimes doing this in rather superfluous contexts] really weird. Is this an established coding style, or just what Alex does!? It's certainly one way to keep to line-width limits, but not the one I would choose - and line limits are violated elsewhere! Extreme example:
+
+                                        self.trigger_fluorescence_image_capture(
+
+                                                                                                                        current_time + timeToWaitInSecs
+
+                                                                                                                        )
+
+Hmm... although I can see that, in the absence of curly brackets for statement blocks, like in C, it helps with readability (in terms of the indenting layout) in cases like this:
+
+if (period != -1
+
+and len(periods) >= (5 + (2 * settings["numExtraRefFrames"]))
+
+and period > 6
+
+and (len(periods) - 1 - settings["numExtraRefFrames"]) > 0
+
+and (periods[len(periods) - 1 - settings["numExtraRefFrames"]]) > 6
+
+                ):
+
+                        periodToUse = periods[len(periods) - 1 - settings["numExtraRefFrames"]]
+
+------------------------------------ **Architecture sketch ------------------------------------**
+
+I am still working on this, but I think there are benefits to having things in their own (concurrent) threads, which in python effectively means they are separate *processes* that must pass messages to each other (or be spawned and return asynchronously). For interfacing with M^2 system, I want a Sync Analyzer process that is driven by [low-level loopback IP?] messages sent back and forth between it and a client. On the RPi, we would also have a Sync Controller process, which would probably be the existing Flask loop, to take care of interfacing with camera/pins hardware. This Sync Controller would be in asynchronous communication with Sync Analyzer, but could also potentially be receiving GUI input from a web client interface.
+
+Messages from Controller to Analyzer would mostly be messages containing frame data, but with other messages to send config information, reset, etc.
```

Comment:
Not entirely sure about the best way to handle streaming frames but: https://www.pyimagesearch.com/2019/04/15/live-video-streaming-over-network-with-opencv-and-imagezmq/

Reply:
TBH, unless Pedro at M^2 has specific other preferences, I am inclined to start with a low-level sockets-based API in which the frames are just sent as binary pixel data (plus metadata such as shape). That's partly just for ease of implementation, to avoid learning yet another framework. But I don't think we'll have "wasted" much time at all, if we later decide to move to something trendier and higher-level.